src/rsz/src/Rebuffer2.cc

```diff
@@ -0,0 +1,2175 @@
+                                       driver_arc,
+                                       slew,
+                                       bnet->cap() + drvr_port_->capacitance(),
+                                       nullptr,
+                                       load_pin_index_map,
+                                       dcalc_ap);
+      rf_delay = dcalc_result.gateDelay();
+      rf_correction = arrival_path->arrival()
+                      - (driver_path->arrival() + dcalc_result.gateDelay());
+      rf_slew = dcalc_result.drvrSlew();
+    } else {
+      rf_delay = 0;
+      rf_correction = 0;
+    }
+
+    if (rf_delay > delay) {
+      delay = rf_delay;
+    }
+    if (rf_correction < correction) {
+      correction = rf_correction;
+    }
+    if (rf_slew > slew) {
+      slew = rf_slew;
+    }
+  }
+
+  return {delay, correction, slew};
+}
+
+bool Rebuffer::loadSlewSatisfactory(LibertyPort* driver, const BnetPtr& bnet)
+{
+  double wire_res, wire_cap;
+  resizer_->wireSignalRC(corner_, wire_res, wire_cap);
+  float r_drvr = driver->driveResistance();
+  float load_slew
+      = (r_drvr + resizer_->dbuToMeters(bnet->maxLoadWireLength()) * wire_res)
+        * bnet->cap() * elmore_skew_factor_;
+
+  return load_slew <= maxSlewMargined(bnet->maxLoadSlew());
+}
+
+Slack Rebuffer::slackAtDriverPin(const BufferedNetPtr& bnet)
+{
+  Delay correction;
+  std::tie(std::ignore, correction, std::ignore) = drvrPinTiming(bnet);
+  return bnet->slack() + correction;
+}
+
+std::optional<Slack> Rebuffer::evaluateOption(const BnetPtr& option,
+                                              // Only used for debug print.
+                                              int index)
+{
+  Delay correction;
+  Slew slew;
+  std::tie(std::ignore, correction, slew) = drvrPinTiming(option);
+  Delay slack = option->slack() + correction;
+
+  if (!loadSlewSatisfactory(drvr_port_, option)) {
+    return {};
+  }
+
+  // Refuse buffering option if we are violating max slew on the driver pin.
+  // There's one exception: If we previously observed satisfactory slew with
+  // an even higher load, ficticiously assume this load satisfies max slew.
+  // This is a precaution against non-mononotic slew vs load data which would
+  // cause inconsistencies in the algorithm.
+  if (slew > drvr_pin_max_slew_ && option->cap() > drvr_load_high_water_mark_) {
+    return {};
+  }
+
+  if (option->cap() > drvr_load_high_water_mark_) {
+    drvr_load_high_water_mark_ = option->cap();
+  }
+
+  debugPrint(logger_,
+             RSZ,
+             "rebuffer",
+             2,
+             "option {:3d}: {:2d} buffers {:.2f} area slack {} cap {}",
+             index,
+             option->bufferCount(),
+             option->area(),
+             delayAsString(slack, this, 3),
+             units_->capacitanceUnit()->asString(option->cap()));
+  return slack;
+}
+
+BnetPtr stripWireOnBnet(BnetPtr ptr)
+{
+  while (ptr->type() == BnetType::wire) {
+    ptr = ptr->ref();
```

Comment:
Are we guaranteed the ptr will never be null (ie we can't reach the end with type wire)?  Same in stripWiresAndBuffersOnBnet

Reply:
Yes, `ref()` is always non-null on type wire or buffer